x-pack/filebeat/input/gcs/job/helper.go

```diff
@@ -0,0 +1,114 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+//go:build !aix
+// +build !aix
+
+package job
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/elastic/beats/v7/libbeat/common/jsontransform"
+	"github.com/elastic/elastic-agent-libs/mapstr"
+)
+
+var (
+	errBodyEmpty       = errors.New("body cannot be empty")
+	errUnsupportedType = errors.New("only JSON objects are accepted")
+)
+
+func httpReadJSON(body io.Reader) (objs []mapstr.M, rawMessages []json.RawMessage, status int, err error) {
+	if body == http.NoBody {
+		return nil, nil, http.StatusNotAcceptable, errBodyEmpty
+	}
+	obj, rawMessage, err := decodeJSON(body)
+	if err != nil {
+		return nil, nil, http.StatusBadRequest, err
+	}
+	return obj, rawMessage, http.StatusOK, err
+}
+
+func decodeJSON(body io.Reader) (objs []mapstr.M, rawMessages []json.RawMessage, err error) {
+	decoder := json.NewDecoder(body)
+	for decoder.More() {
+		var raw json.RawMessage
+		if err := decoder.Decode(&raw); err != nil {
+			if err == io.EOF { //nolint:errorlint // This will never be a wrapped error.
+				break
+			}
+			return nil, nil, fmt.Errorf("malformed JSON object at stream position %d: %w", decoder.InputOffset(), err)
+		}
+
+		var obj interface{}
+		if err := newJSONDecoder(bytes.NewReader(raw)).Decode(&obj); err != nil {
+			return nil, nil, fmt.Errorf("malformed JSON object at stream position %d: %w", decoder.InputOffset(), err)
+		}
+		switch v := obj.(type) {
+		case map[string]interface{}:
+			objs = append(objs, v)
+			rawMessages = append(rawMessages, raw)
+		case []interface{}:
+			nobjs, nrawMessages, err := decodeJSONArray(bytes.NewReader(raw))
+			if err != nil {
+				return nil, nil, fmt.Errorf("recursive error %d: %w", decoder.InputOffset(), err)
+			}
+			objs = append(objs, nobjs...)
+			rawMessages = append(rawMessages, nrawMessages...)
+		default:
+			return nil, nil, errUnsupportedType
+		}
+	}
+	for i := range objs {
+		jsontransform.TransformNumbers(objs[i])
+	}
+	return objs, rawMessages, nil
+}
+
+func decodeJSONArray(raw *bytes.Reader) (objs []mapstr.M, rawMessages []json.RawMessage, err error) {
```

Comment:
Same, and I'm wondering why we need to have two function in this recurrence; JSON doesn't have a complex recursive structure so one function should be enough.